SO/Ficha2.md

```diff
@@ -0,0 +1,290 @@
+terminado.
+
+O `waitpid` por outro lado, permite um ter mais controlo sobre quando retornar.
+
+O parâmetro `pid` pode tomar valores que não `pid`s, mas para além de `-1` (que
+significa "qualquer processo"), estes saem fora da matéria da cadeira
+[<sup>\[2\]</sup>](processGroups).
+
+Ao parâmetro `options` pode ser passado `0` para não ativar nenhuma opção ou
+as opções especificadas na *man page*, das quais a mais útil é `WNOHANG` que
+permite ao `waitpid` não bloquear caso nenhum processo filho tenha terminado.
+
+O que acontece ao processo filho se não fizermos wait?
+
+Enquanto o processo pai existir os processos filhos que já terminaram são
+designados de *zombies*, o sistema operativo mantém apenas a informação
+necessária para que o processo possa ser *waited for* mais tarde. Mesmo assim,
+enquanto o *zombie* existir uma entrada na tabela de processos continua a
+existir e se esta tabela encher não é possível criar novos processos.
+
+No caso de o pai terminar, normalmente, sem ter feito *wait* de todos os seus
+filhos, estes tornam se órfãos e passam a ser filhos do processo `init` que tem
+sempre `pid` 1.
+
+O `init` é o primeiro processo que arranca sempre que o sistema operativo
+inicia, tem sempre `pid` 1 e é o "pai de todos". Ele encarrega-se de
+periodicamente chamar o `wait` para todos os seus filhos de forma a evitar ter
+*zombies*, desta forma também serve como um *fall back plan* para processos
+órfãos não se tornarem *zombies* mais tarde.
+
+## Exit status
+
+Quando um programa termina retorna uma valor entre `0` e `255`, e o pai desse
+processo pode verificar esse número usando as macros `WIFEXITED` e
+`WEXITSTATUS`. Estas tem de ser usadas em conjunto para evitar *undefined
+behaviour*, sendo que `WEXITSTATUS` só pode ser usada caso `WIFEXITED` retornar
+`1` (`true`).
+
+```c
+int status;
+wait(&status);
+if(WIFEXITED(status)) {
+    puts("Child exited normally");
+    int exit_status = WEXITSTATUS(status);
+    printf("Exit status: %d\n", status);
+}
+```
+A razão para a ter de ser verificar se o processo terminou normalmente é que
+este pode ter terminado por receber um sinal, por exemplo, `SIGSEGV`
+(segmentation fault), e neste caso o valor que vem em `status` não é um exit
+status valido.
+
+## Exemplos
+
+### Procurar em paralelo.
+```c
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <stdlib.h>
+#include <assert.h>
+
+#define SIZE 1000 * 1024 * 1024
+#define LINES 2
+
+static char BIG_DATA[LINES][SIZE] = {0};
+
+int child(size_t line) {
+    for(size_t j = 0; j < SIZE; j++) {
+        if (BIG_DATA[line][j] == 'x') {
+            printf("Found it at: (%zu,%zu)\n", line, j);
+            return EXIT_SUCCESS;
+        }
+    }
+    return EXIT_FAILURE;
+}
+
+int main(int argc, char const* argv[]) {
+    assert(argc > 2); // Can't bother handling lack of args
+    // Putting the 'x' in the array to find it.
+    BIG_DATA[atoi(argv[1]) % 2][atoi(argv[2])] = 'x';
```

Comment:
`atoi` is dangerous, use `strtoi` instead

Reply:
I know, but I can't bother and argument parsing is not the focus of the exercise.

`strtoi` requires way too much code for this.